Add ability to sign "arbitrary" CSRs

nielslaukens opened the pull request:
I was looking for a Let's Encrypt client to integrate in to an existing automation tool. Your `letsencrypt.sh` implementation came closest. However, it still does too much.

This pull requests splits out the core functionality (sign a CSR) from the nice-to-have features (generate a CSR from the `domains.txt` file), and exposes this functionality from the command line.

The newly added `--signcsr` command does exactly what its name implies: it extracts the domain names from the CSR (SANs or CN), requests and executes the challenge for each one, requests the CRT, and outputs that to stdout (normal output is redirected to stderr for this command).

rossnick commented:
I was trying this out, and I got : 

sed: -e expression #1, char 23: invalid reference \1 on `s' command's RHS

 + Requesting certificate...
  + ERROR: An error occurred while sending post-request to https://acme-v01.api.letsencrypt.org/acme/new-cert (Status 403)

nielslaukens commented:
Hi rossnick,
Thank you for trying this out. Can you provide me with the exact command line that you used, and the input CSR file?

rossnick commented:
The command was :  /etc/httpd/ssl/letsencrypt.sh -s file.csr -f /etc/httpd/ssl/config.sh Can I send the csr in private ? By the way, the CSR is in sha1, it's used by a propriatary system that doesn't support creating csr in sha2. That's the reason I am interrested in testing this...

nielslaukens commented:
there is no reason to send the CSR in private (Let's Encrypt will publish all details anyway if they issue a CRT for it), but if you insist, you can mail it to <niels@dest-unreach.be>.

rossnick commented:
You're right...

-----BEGIN CERTIFICATE REQUEST-----
MIIE5jCCAs4CAQAwgaAxCzAJBgNVBAYTAkNBMQ8wDQYDVQQIEwZRdWViZWMxDzAN
BgNVBAcTBlF1ZWJlYzEWMBQGA1UEChMNQ3liZXJDYXQgaW5jLjELMAkGA1UECxMC
SVQxJTAjBgNVBAMTHG1haWwuY29tbW9ud2VhbHRocGx5d29vZC5jb20xIzAhBgkq
hkiG9w0BCQEWFHNzbGFkbWluQGN5YmVyY2F0LmNhMIICIjANBgkqhkiG9w0BAQEF
AAOCAg8AMIICCgKCAgEAuwLQL6PQ+92/AHTxOdNz4fveWZt2OqBIkdcGEs5r2arK
sMW9bN+cJ46CySKOaMwZWnFgsBeCLa/KMipVc3dFlPpg698mOVZRGCCtKwt3Iika
Y74tY6czeUofg9/dnRXheUplwXsKv9t6/t1ab9hr+UVU4PI/y0FDH9+LuP2qA41g
xPtuFHvrpi5K0t9PeRKA6n19GY9VxbGYd1d0Rbc9HovwEbTrXSghJnR7FzHm3d0C
LAPoUyrZsGNK4j7ebphtV9eMVsXiOYL9EquekQXAVB+S3oWbpJZs6qwK2p+/sUz0
yl98k+Kmo3G9gUygnJ+hhXqjqv7IKtmVkI6X2IdDRwutzHORPQIXjSHp2K0yiMuV
MVsu/WBLYx1OAoNRdAYkB880D95TzCc0McUc+YM/+5TJKaoBCid0yeQn2ZmvninD
MlzKMaalF5DLv7Yh+cfs31bq3rYKohY4jq9Wt/0u52lIA303XuwVUbhFvvi+O4zk
3PpWnyWD6xWq2dfYgO3YrwGUa4TyZiNJR44ZS+B+p7YOP5yZ9xhbC9XyzlkQCD5h
dzEDYjcgzHuz2u/38eHxaGjO6fcpuOS/uYtvRi/KkixqaE8QVxylOYpynQM6QLhw
qIqcQjQVEM28Of3Vcoym1bS65ihJl/wn1/6X40u+BDF6Yx7aVuID4IlNv71fhaEC
AwEAAaAAMA0GCSqGSIb3DQEBBQUAA4ICAQAU7L0YyfOkJQlW6bCP74vjlnhfza/9
WY1R8LRLi1zsgwao+d+2K5IGPTGLrtNdWLcccprZobeJoUkpNj2shKnbhyyLglfp
gqMmE3PsCHq+8y9Rk9N5nXeDuUll5sD4XWxoPwdES9pYmJ3vAtedx9dxaRotEB4I
2SzAd5dQgKcZfLImUSADZCDuL/K1FH638c6gUG1jtRs9Ylb01ifM+ALrEVVFvd6C
UtNI/uXpG94jAtK8AkHgYbTKLzUfM96nqSom8T8+28T4DmdobrAEgYstIEeY674A
vAlY2IgvtOivavde/Pfa8n15o37rSFvEVSrHpgD6EgL7tM8orsngwXYBY4L1a+oj
YuvSfIGHDYSgKaFiqxyIYBj+E0RggTxI0Rtu/mARPlsgBdJ8Lj1gYuX9+OFxTHGb
tY1ysOpxxeKwJEStp+//TttmLHcCvfLsNkYjHXIYzTeqJzxrIMiwvWb8alR17/AH
RjvJyvp5cOMzrWXt4YeMkYG1XhyOcqmnKKXY9FTdxNWBFcC6DBAd31LG2j3IsQ4B
tqZ32rnND26kmSKaCe5O+xmfTWKKT3SvY0fVrbw/HNUIiXX2M6MjFR8DiSEz7P99
JuNXypdIy82fFTmEUlmvzHYJvVr3PWDlMg9WtE6gyS9a21zsTZax9VUEwM5U/QpE
pL//s+RsJNyBxg==
-----END CERTIFICATE REQUEST-----

nielslaukens commented:
Does the current version work (better)?

rossnick commented:
Yes !
It worked. It would be nice if it saved the file into the certs/domain/ directory and making of the chains, like when signing certs.

nielslaukens commented:
I've rebased the patch in to the offending commit, to make this pull request easier to review. (And I've rebased it on to the new `master`-branch as well)

nielslaukens commented:
rossnick: I don't consider making chains and creating arbitrary directory structures as 'minimal functionality'. Also, for my particular use case, I need to put the certs in a different directory structure. You can always save the certificate by redirecting stdout: `./letsencrypt.sh --signcsr bla.csr > certs/bla/cert.crt`

rossnick commented:
I tested the generated cert into our old product and was able to confirm that all is working great. I'd love to see this added to the main project.

lukas2511 commented:
Thanks, this is something I actually wanted to add to this project at some point, so this is great!

I didn't yet have time to look at the code, but if it's fine I'll merge it. I'll try to do that tomorrow.

nielslaukens commented:
@lukas2511 I've tried to make the commits as easy to review as possible. However, the first commit (splitting the challange from the CSR generation) gives a horrible to read diff. It's probably easier if you just open up both versions side-by-side and do the diff by hand.
